synapse/cortex.py

```diff
@@ -6179,6 +6199,557 @@ async def iterTagPropRows(self, layriden, tag, prop, form=None, stortype=None, s
+
+            if vault:
+                self._reqEasyPerm(vault, user, s_cell.PERM_READ)
+
+            return vault
+
+        # If caller specified a scope, return that vault if it exists
+        if scope is not None:
+            vault = _getVault(scope)
+            if vault:
+                return vault.get('data', {})
+
+            return None
+
+        # Now try the default vault
+        byts = self.slab.get(vtype.encode(), db=self.vaultdefaultsdb)
+        if byts and (default := s_msgpack.un(byts)):
+            vault = _getVault(default)
+            if vault:
+                return vault.get('data', {})
+
+        # Finally, try the user, role, and global vaults in order
+        for _scope in ('user', 'role', 'global'):
+            vault = _getVault(_scope)
+            if vault:
+                return vault.get('data', {})
+
+        return None
+
+    async def addVault(self, name, vtype, scope, iden, data, user=None):
+        '''
+        Create a new vault.
+
+        Permissions: vaults.<scope>.add
+
+        Args:
+            name (str): Then name of the new vault.
+            vtype (str): Type of the vault to create.
+            iden (str): User iden for user scoped vaults. Role iden for role
+                        scoped vaults. None for global vaults.
+            data (object): Vault data. Must conform to the schema specified by `vtype`.
+            user (Optional[HiveUser]): User trying to create the vault.
+
+        Raises:
+            synapse.exc.BadArg:
+                - `vtype` is not a valid vault type
+                - `scope` is not one of 'user', 'role', 'global'
+
+            synapse.exc.AuthDeny: `user` does not have permission to create vault.
+
+        Returns: iden of new vault
+        '''
+        if scope not in (None, 'user', 'role', 'global'):
+            raise s_exc.BadArg(f'Invalid scope provided: {scope}')
+
+        try:
+            s_msgpack.en(data)
+        except s_exc.NotMsgpackSafe as exc:
+            raise s_exc.BadArg(mesg=f'Vault data must be msgpack safe.') from None
+
+        if iden is not None and scope not in ('user', 'role'):
+            raise s_exc.BadArg('Scope must be user/role when providing iden')
+
+        if user is None:
+            user = self.auth.rootuser
+
+        # Iden is a randomly generated guid to uniquely identify this vault
+        viden = s_common.guid()
+
+        vault = {
+            # Name is a combination of '<scope>:<ident>:<vtype>'
+            'name': name,
+
+            # Type is one of vault types from addVaultType()
+            'type': vtype,
+
+            # Scope is one of {user, role, global}
+            'scope': scope,
+
+            # Ident is the iden of the user/role that created this vault
+            'ident': iden,
+
+            # Data is an opaque thing defined by the type. The only requirement
+            # is that we can msgpack it
+            'data': data,
```

Comment:
The existence of the setVaultData() API seems to imply that this must be a dictionary?